fix(rpc): Remove additional encoding for rpc utxo scripts

SatwikPrabhu opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Pull request titles must use the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) format -->

## Issue being fixed or feature implemented

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please provide a link to the issue -->

There was a consistency in the format of the `script` field in `RpcUtxo` and `DbUtxo`. This would be potentially confusing down the line when using a grpc endpoint that uses `RpcUtxo`.



## What was done?

<!--- Describe your changes in detail -->

Now the format is consistent, so both structs store the scriptpubkey without the the length prefix. i.e. a p2tr script just starts with `OP_1`, not `0x22`.

To maintain backward compatibility, the method `TryFrom<RpcUtxo> for DbUtxo` will first interpret the RpcUtxo script with the new format, and if the result is not a valid p2tr script, fallback to interpreting it with the prefix. 

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

- Unit and integration tests

## Breaking Changes

<!--- Please describe any breaking changes your code introduces and verify that -->
<!--- the title includes "!" following the conventional commit type (e.g. "feat!: ..."-->

- None

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

-   [x] I have added or updated relevant unit/integration/functional/e2e tests
-   [ ] I have tested my changes running a local network, and they work as expected
-   [x] I have performed a self-review
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
-   [x] I have made corresponding changes to the documentation if needed
-   [x] I have added assignees and reviewers to this pull request
-   [x] I have added the PR to the project board or linked it to an issue from the board


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Improved compatibility when reading/writing Bitcoin scripts: supports both raw Taproot and legacy-encoded formats, reducing deserialization issues.
- Improvements
  - Added warning logs when legacy script formats are detected to aid operators.
- Bug Fixes
  - More reliable UTXO conversions with robust round-trip handling across script formats.
- Tests
  - Updated test suites to use proper scriptPubKey generation and direct byte encoding.
  - Added tests covering fallback parsing and round-trip conversions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SatwikPrabhu committed:
Fix(rpc): Remove additional encoding for rpc utxo scripts

scottmillner commented:
will review when pipeline is green

shumkov commented:
@SatwikPrabhu please resove conflicts

DariusParvin commented:
I'm quite weary about making this change for the following reasons:
- `RpcUtxo` is used in new_consensus_checkpoint
- the format of scripts in the db is quite critical, so even if there is some inefficiency, it's safer to just leave it as is.
- there's no unit test for the type conversions

I'm guessing from the PR description that the expected behavior was a different encoding of the script when calling the `GetAllUtxos` struct. I don't believe this is being used right now, and was this probably came up while developing the sweep feature.

I think changing the conversion of the `From` conversion, but not the other direction seems problematic, because then there would be an inconsistent definition of what the `script` field in the struct should be. Maybe this fix is actually making it so that the types are consistent, but without unit tests it's hard to tell. 

I think the next best step here (that I will work on) is to add a unit test to do a round conversion, from db utxo -> rpc utxo and back -> db utxo, and verify that nothing was lost. If this change makes that correct, then we can merge, but if not, then maybe it would be better to just deprecate this struct and use a new definition when it's needed?

DariusParvin committed:
refactor: cargo fmt

DariusParvin committed:
test(btc-server): add unit test for db rpc utxo conversion

rwlockbg commented on bin/btc-server/src/util.rs:
```diff
@@ -748,7 +748,7 @@ mod tests {
         assert_eq!(
             res.unwrap_err(),
-            ValidatePSBTError::FeeSanityCheck(Amount::from_sat(2346), Amount::ZERO)
+            ValidatePSBTError::FeeSanityCheck(Amount::from_sat(2332), Amount::ZERO)
```
Is this some sort of a magic number here ? Could we add some explanation as to how it is derived ?

DariusParvin replied:
yeah this is from an old test before we had more proper fee calculation methods. By the way, I think this whole PR is probably not worth it. I'm going to comment on the PR and then close it.

DariusParvin replied:
@rwlockbg  that value was the fee rate of the psbt, but that is outside the scope of the unit test. It was just included to match the error message that was returned when the fee is too high. There are other tests for calculating the exact fee rate. 

I've updated this test so that the assertion uses `matches!` on the error type returned, so that it doesnt look like there's a magic number there

DariusParvin commented:
In addition to the other comment I wrote above, i just think this change is introducing unnecessary risk for the sake of addressing a small inconsistency between the RpcUtxo and DbUtxo. 

This is effectively changing the grpc contract of what data should be passed in for the scriptPubkey field of the RpcUtxo. That's why a lot of the test inputs have to be updated in this PR, but there are some scenarios i can think of where this happens in prod:
- The btc server and reth server are running different versions
- we miss some part of the code that should have been updated, or we try to revert some changes, without considering how RpcUtxo had changed.

We _could_ add some extra validation or build in some backwards compatibility, but again I think it's not worth introducing that complexity or risk.

DariusParvin commented:
On second thought I think i'll just add some safety checks in the conversion, so that we dont end up with invalid scripts in the db. The pro of doing this now is that we avoid running into the same issue in the future

DariusParvin committed:
fix(btc-server): add backwards compatibility for RpcUtxo script format change

DariusParvin committed:
test(btc-server): remove exact fee amount from psbt validation unit test

The exact fee is not what's being tested in the unit test, so it should be removed to make the scope of the test clearer
